### PR TITLE
[JIT] Add property support for ScriptModules

### PR DESCRIPTION
**Pull request by SplitInfinity**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42390 [JIT] Add property support for ScriptModules**

**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

### CODE REVIEW

File: torch/jit/_recursive.py [hunk 1]
```diff
@@ -239,14 +252,6 @@ def infer_type(name, item):
                     "to a TorchScript type.)").format(torch.typename(type(value)))
             concrete_type_builder.add_failed_attribute(name, hint)
 
-    # Add @property methods as failed attributes, to give a better error message.
-    for name, value in type(nn_module).__dict__.items():
-        if isinstance(value, property):
-            hint = ("\n(This attribute exists on the Python module, but it's an @property "
-                    "method. @property methods are not yet supported in TorchScript. "
-                    "Please file a feature request on Github)")
-            concrete_type_builder.add_failed_attribute(name, hint)
```
Comment (SplitInfinity) [inline torch/jit/_recursive.py:248]: I wasn't quite sure what to do here. Properties are really just syntactic sugar for setter and getter methods. Do they belong on the concrete type? If yes, how do we represent them? If not, what happens if two modules have the same underlying data attributes but different property getters/setters for them?

File: torch/csrc/jit/python/python_tree_views.cpp [hunk 2]
```diff
@@ -162,7 +162,18 @@ void initTreeViewBindings(PyObject* module) {
...
+      .def("name", [](const Property& property) { return property.name(); })
+      .def(
+          "getter_name",
+          [](const Property& property) { return property.getter().name(); })
+      .def("setter_name", [](const Property& property) {
+        if (property.setter().present()) {
+          return property.setter().get().name();
+        }
+
+        return Ident::create(property.range(), "");
```
Comment (eellison) [inline torch/csrc/jit/python/python_tree_views.cpp:175]: Can we use None here instead ? empty string as null value could lead to some confusing debugging

File: torch/jit/_recursive.py [hunk 3]
```diff
@@ -29,6 +31,16 @@
     "dump_patches",
 ]
 
+ignored_properties = [
+    # Temporary fix for RNN module property named 'all_weights' being scripted
+    "all_weights",
```
Comment (eellison) [inline torch/jit/_recursive.py:36]: Can we switch this instead to a list of (Python Class, [ignored_properties]) tuples, and we ignore a property if the scripted class is a subclass of our registered class ? right now we're ignoring these names indiscriminately

File: torch/jit/_recursive.py (hunk 1 above)
Comment (eellison) [reply torch/jit/_recursive.py:248]: So, concrete type already returns false if two classes differ: pyClass_.is(other.pyClass_). As long as we only take properties from methods, and not from attributes, we don't need to do any additional work here. If we do take properties as attributes than we'll need to add stuff to the ConcreteType. 

(i think it's fine to not take properties that are passed in as attributes, no one really does that)

File: torch/jit/_recursive.py [hunk 4]
```diff
@@ -407,14 +419,20 @@ def init_fn(script_module):
...
         # Add the methods to the script_module directly. This ensures they will
         # be found first when `name` is looked up (as opposed to the stubs or
         # nn.Module.forward)
         script_module.__dict__[name] = script_method
 
 
+    for property_stub in property_stubs:
+        property_name = property_stub.def_.name().name
+        fget = cpp_module._get_method(property_stub.def_.getter_name().name)
+        fset = cpp_module._get_method(property_stub.def_.setter_name().name)
```
Comment (eellison) [inline torch/jit/_recursive.py:433]: don't we need to handle a setter not begin set ?

**Review by eellison — COMMENTED** [review]
Looks good, a few small comments

File: torch/jit/_recursive.py (hunk 4 above)
Comment (SplitInfinity) [reply torch/jit/_recursive.py]: Good catch, fixed.

File: torch/jit/_recursive.py (hunk 3 above)
Comment (SplitInfinity) [reply torch/jit/_recursive.py]: Fixed. However, I had to use `torch.nn.Module` as the class for all of the `RecursiveScriptModule` properties that we need to ignore because of circular import problems.

File: torch/csrc/jit/python/python_tree_views.cpp (hunk 2 above)
Comment (SplitInfinity) [reply torch/csrc/jit/python/python_tree_views.cpp:175]: `None` as a string literal? I tried to use `py::none()` but that didn't work because of what I think is a problem inferring the return type of this lambda. Is there an example you can point me to?

Comment (eellison) [reply torch/csrc/jit/python/python_tree_views.cpp:175]: you can return None with c10::optional, and cast both returns to `c10::optional<Ident>` so that the return type gets properly inferred

Comment (SplitInfinity) [reply torch/csrc/jit/python/python_tree_views.cpp]: Fixed.

File: torch/jit/_recursive.py [hunk 5]
```diff
@@ -407,14 +424,21 @@ def init_fn(script_module):
...
         # nn.Module.forward)
         script_module.__dict__[name] = script_method
 
 
+    for property_stub in property_stubs:
+        property_name = property_stub.def_.name().name
+        fget = cpp_module._get_method(property_stub.def_.getter_name().name)
+        setter_name = property_stub.def_.setter_name()
+        fset = cpp_module._get_method(setter_name.name) if setter_name else None
+        script_module.__dict__[property_name] = property(property_name, fget, fset)
```
Comment (eellison) [inline torch/jit/_recursive.py:440]: just for my own understanding, why was this needed ? maybe add a comment

**Review by eellison — APPROVED** [review]
(`-,-,
                                   ('(_,( )
                                    _   `_'
                                 __|_|__|_|_
                               _|___________|__
                              |o o o o o o o o/  
                             ~'`~'`~'`~'`~'`~'`~

File: torch/jit/_recursive.py [hunk 6]
```diff
@@ -548,6 +580,47 @@ def ignore_overloaded(method_name):
         stubs.append(make_stub_from_method(nn_module, method))
     return overload_stubs + stubs
 
+
+def should_compile_property(cls, property_name):
```
Comment (eellison) [inline torch/jit/_recursive.py:584]: Now that we've added ignore, can we just add ignores on the modules themselves instead of maintaining a special list?

File: torch/jit/_recursive.py [hunk 7]
```diff
@@ -371,16 +371,22 @@ def init_fn(script_module):
...
                 bound_method = unbound_function.__get__(script_module)
                 setattr(script_module, name, bound_method)
+            elif isinstance(item, property) and _jit_internal.is_ignored_fn(item.fget):
+                unbound_getter = getattr(type(nn_module), item.fget.__name__)
+                bound_getter = unbound_getter.__get__(script_module)
+                unbound_setter = getattr(type(nn_module), item.fset.__name__) if item.fset else None
+                bound_setter = unbound_setter.__get__(script_module) if unbound_setter else None
+                setattr(script_moudle, name, property(name, bound_getter, bound_setter))
+            else:
+                pass
```
Comment (eellison) [inline torch/jit/_recursive.py:389]:
```
else:
     pass
```
is dead code

File: torch/jit/_recursive.py [hunk 8]
```diff
@@ -371,16 +371,22 @@ def init_fn(script_module):
...
             item = getattr(nn_module, name, None)
-            if not inspect.ismethod(item):
-                continue
-            if _jit_internal.is_ignored_fn(item):
+            if inspect.ismethod(item) and _jit_internal.is_ignored_fn(item):
                 unbound_function = getattr(type(nn_module), name)
                 bound_method = unbound_function.__get__(script_module)
                 setattr(script_module, name, bound_method)
+            elif isinstance(item, property) and _jit_internal.is_ignored_fn(item.fget):
+                unbound_getter = getattr(type(nn_module), item.fget.__name__)
```
Comment (eellison) [inline torch/jit/_recursive.py:383]: why are we doing anything if `_jit_internal.is_ignored_fn(item.fget` ?

**Review by eellison — COMMENTED** [review]
a few questions

File: torch/jit/_recursive.py (hunk 6 above)
Comment (SplitInfinity) [reply torch/jit/_recursive.py:584]: Yeah good point.

File: torch/jit/_recursive.py (hunk 8 above)
Comment (SplitInfinity) [reply torch/jit/_recursive.py:383]: Don't we still need to attach the property to the scripted module? Just as we do with ignored functions?

File: torch/_jit_internal.py [hunk 9]
```diff
@@ -463,6 +463,15 @@ def forward(self, x):
         fn._torchscript_modifier = FunctionModifiers.IGNORE
         return fn
 
+    if isinstance(drop, property):
+        prop = drop
+        setattr(prop.fget, "_torchscript_modifier", FunctionModifiers.IGNORE)
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/_jit_internal.py:468]: I had to use setattr due to mypy failures. A file for which mypy is enabled is typechecked, and mypy is instructed to recursive into this file to continue typechecking (I think). `# type : ignore` didn't work.

[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/_jit_internal.py:468]: The specific error in question was about setting an attr on a `Callable` that doesn't exist.

File: torch/jit/_recursive.py (hunk 8 above)
Comment (eellison) [reply torch/jit/_recursive.py:383]: yea, good point

File: torch/jit/_recursive.py (hunk 6 above)
Comment (SplitInfinity) [reply torch/jit/_recursive.py]: Sigh actually we can't do this because you can't dynamically add properties to an instance. You have to add them to the class. :'(

**Comment by SplitInfinity** [comment]
Ping @eellison 

I know you already approved this but the remaining items are more contentious than I thought so I just want to close them out.

File: torch/_jit_internal.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/_jit_internal.py:468]: Oh but using `setattr` causes a lint failure :(

**Review by eellison — APPROVED** [review]
LGTM
